http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeV2ServiceInterp.scala

```diff
@@ -179,27 +202,79 @@ class AzureServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
 
-      _ <- context.span.traverse(s => F.delay(s.addAnnotation("Done auth call for delete azure runtime permission")))
+      _ <- ctx.span.traverse(s => F.delay(s.addAnnotation("Done auth call for delete azure runtime permission")))
       _ <- F
         .raiseError[Unit](RuntimeNotFoundException(cloudContext, runtimeName, "permission denied"))
         .whenA(!hasPermission)
 
-      _ <- clusterQuery.markPendingDeletion(runtime.id, context.now).transaction
+      _ <- clusterQuery.markPendingDeletion(runtime.id, ctx.now).transaction
 
       // For now, azure disk life cycle is tied to vm life cycle and incompatible with disk routes
-      _ <- persistentDiskQuery.markPendingDeletion(diskId, context.now).transaction
+      _ <- persistentDiskQuery.markPendingDeletion(diskId, ctx.now).transaction
 
       _ <- publisherQueue.offer(
-        DeleteAzureRuntimeMessage(runtime.id, Some(diskId), workspaceId, wmsResourceId, Some(context.traceId))
+        DeleteAzureRuntimeMessage(runtime.id, Some(diskId), workspaceId, wmsResourceId, Some(ctx.traceId))
       )
     } yield ()
 
+  override def listRuntimes(
+    userInfo: UserInfo,
+    workspaceId: Option[WorkspaceId],
+    cloudProvider: Option[CloudProvider],
+    params: Map[String, String]
+  )(implicit as: Ask[F, AppContext]): F[Vector[ListRuntimeResponse2]] =
+    for {
+      paramMap <- F.fromEither(processListParameters(params))
+
+      runtimes <- RuntimeServiceDbQueries.listRuntimes(paramMap._1, paramMap._2, workspaceId, cloudProvider).transaction
+
+      (runtimesUserIsCreator, runtimesUserIsNotCreator) = runtimes.partition(_.auditInfo.creator == userInfo.userEmail)
+
+      // Here, we check if backleo has updated the runtime sam id with the wsm resource's UUID via type conversion
+      // If it has, we can then use the samResourceId of the runtime (which is the same as the wsm resource id) for permission lookup
+      wsmControlledResourceSamIds = runtimesUserIsNotCreator
+        .flatMap { r =>
+          for {
+            uuid <- Either.catchNonFatal(UUID.fromString(r.samResource.resourceId)) match {
+              case Right(id) => List(id)
+              case Left(_)   => List.empty
+            }
+          } yield WsmResourceSamResourceId(WsmControlledResourceId(uuid))
+        }
+
+      samVisibleWsmControlledResourceSamIds <- NonEmptyList
+        .fromList(wsmControlledResourceSamIds)
+        .fold[F[List[WsmResourceSamResourceId]]](F.pure(List.empty)) { ids =>
+          authProvider.filterUserVisible(ids, userInfo)
+        }
+
+      // We must also check the RuntimeSamResourceId in sam to support already existing and newly created google runtimes
+      samVisibleRuntimeSamResourceIds <- NonEmptyList
+        .fromList(runtimesUserIsNotCreator.map(_.samResource))
+        .fold[F[List[RuntimeSamResourceId]]](F.pure(List.empty)) { ids =>
+          authProvider.filterUserVisible(ids, userInfo)
```

Comment:
does this still have equivalent of project level fallback? It doesn't like this logic do anything special if the caller is workspace owner?

Reply:
We do check if the caller is the creator of the runtime, but I'm not sure exactly what a 'project fallback' entails from the code. Are you saying we also want to check if the user is he owner of the workspace?

Reply:
sry just saw the comment,

>So the corollary would be to check if the user is he owner of the workspace in addition to the sam runtime permission we are checking now for this use case?

yes!

Reply:
cc @Qi77Qi